### PR TITLE
[Sources] Auto expand path of selected source

### PR DESCRIPTION
**Pull request by AnshulMalik**
### Summary of Changes

* Expand all nodes in the path of the selected source in sources tree

- [x] write tests for this

### Screenshots/Videos (OPTIONAL)
auto-expand

### CODE REVIEW

File: src/components/shared/ManagedTree.js [hunk 1]
```diff
@@ -121,15 +124,24 @@ class ManagedTree extends Component<Props, State> {
...
+    const reversedItems = highlightItems.reverse();
+    let index = reversedItems.findIndex(
+      item => !expanded.has(this.props.getPath(item))
+    );
+
+    if (this.props.autoExpandOnHighlight) {
+      while (index <= reversedItems.length - 2) {
+        this.setExpanded(reversedItems[index], true, false);
+        index++;
+      }
```
Comment (jasonLaster) [inline src/components/shared/ManagedTree.js]: hmm, can you discuss what this does?

File: src/reducers/source-tree.js [hunk 2]
```diff
@@ -14,13 +14,14 @@ import makeRecord from "../utils/makeRecord";
...
 export type SourceTreeState = {
-  expanded: any
+  expanded: Expanded
 };
 
-export function InitialState(): Record<SourceTreeState> {
+function InitialState(): Record<SourceTreeState> {
   return makeRecord({
-    expanded: null
+    expanded: new Set()
```
Comment (jasonLaster) [inline src/reducers/source-tree.js]: nice

File: src/components/shared/ManagedTree.js (hunk 1 above)
Comment (AnshulMalik) [reply src/components/shared/ManagedTree.js]: I have added a comment to make it more clear

**Comment by darkwing** [comment]
Kicked Travis, let's see if we pass now :)

**Comment by darkwing** [comment]
We'll want a mochitest or two for this but I looooove this functionality.  Keep up the great work @AnshulMalik !

**Comment by AnshulMalik** [comment]
I am confused with what's going on in `browser_dbg-asm.js`

```diff
diff --git a/src/test/mochitest/head.js b/src/test/mochitest/head.js
index 873d1ada..f8dd68ce 100644
--- a/src/test/mochitest/head.js
+++ b/src/test/mochitest/head.js
@@ -212,8 +212,12 @@ function waitForSources(dbg, ...sources) {
       }

       if (!sourceExists(store.getState())) {
+        info(`WAITING FOR ${url}`);
         return waitForState(dbg, sourceExists, `source ${url}`);
       }
+
+      info('RESOLVING DIRECTLY, ALREADY EXISTS');
+      return true;
     })
   );
 }
```

It says, wait for two sources, but the sources are already loaded, `Promise.all()` should resolve immediately (I see `RESOLVING DIRECTLY, ALREADY EXISTS` twice for the two sources).
But it's not resolving. Not sure what's the problem here

**Comment by darkwing** [comment]
Been trying to figure out the issue here but having no luck so far.  Oddly enough, the tests pass for me locally.

**Comment by jasonLaster** [comment]
probably unrelated, but a rebase is always healthy

File: src/components/shared/ManagedTree.js [hunk 3]
```diff
@@ -121,15 +124,25 @@ class ManagedTree extends Component<Props, State> {
...
-      const index = highlightItems
-        .reverse()
-        .findIndex(item => !expanded.has(this.props.getPath(item)));
-      this.focusItem(highlightItems[index]);
+      return this.focusItem(highlightItems[0]);
+    }
+
+    let index = highlightItems.length - 2;
+
+    if (this.props.autoExpandOnHighlight) {
```
Comment (darkwing) [inline src/components/shared/ManagedTree.js]: I turned `autoExpandOnHighlight` off and it didn't change the fact that the test failed.

File: src/components/shared/ManagedTree.js [hunk 4]
```diff
@@ -121,15 +124,25 @@ class ManagedTree extends Component<Props, State> {
...
-      // Look at folders starting from the top-level until finds a
-      // closed folder and highlights this folder
-      const index = highlightItems
-        .reverse()
-        .findIndex(item => !expanded.has(this.props.getPath(item)));
-      this.focusItem(highlightItems[index]);
+      return this.focusItem(highlightItems[0]);
+    }
+
+    let index = highlightItems.length - 2;
```
Comment (darkwing) [inline src/components/shared/ManagedTree.js]: We should comment this; the `-2` value is a mystery

**Pull request merged**